Non-blocking stdio isn't working correctly

**Describe the bug**

Non-blocking stdio isn't working correctly.

**To Reproduce**

Use [`gotip`](https://pkg.go.dev/golang.org/dl/gotip) that includes at least https://github.com/golang/go/commit/13529cc5f443ef4e242da3716daa0032aa8d34f2 (note that this will be part of the Go 1.21 release in ~August).

If I compile the following module:

```go
package main

import (
	"io"
	"os"
	"time"
)

func main() {
	go func() {
		time.Sleep(1 * time.Second)
		os.Stdout.WriteString("waiting for stdin...\n")
	}()

	b, err := io.ReadAll(os.Stdin)
	if err != nil {
		panic(err)
	}
	os.Stdout.Write(b)
	return
}
```

using:

```
$ GOOS=wasip1 GOARCH=wasm gotip build -o stdio.wasm stdio.go
```

and then run the script with wazero, it returns errno=28 (EINVAL) from `poll_oneoff` which causes Go to crash:

```
$ (sleep 3; echo foo) | wazero run stdio.wasm 
waiting for stdin...
errno= 28  len(pollsubs)= 1
fatal error: poll_oneoff failed

runtime stack:
runtime.throw({0x17c29, 0x12})
	/Users/chris/sdk/gotip/src/runtime/panic.go:1077 +0x3 fp=0xeb778 sp=0xeb750 pc=0x12250003
runtime.netpoll(0xffffffffffffffff)
	/Users/chris/sdk/gotip/src/runtime/netpoll_wasip1.go:218 +0x45 fp=0xeb7e8 sp=0xeb778 pc=0x11f30045
```

**Expected behavior**

[wasi-go](https://github.com/stealthrocket/wasi-go) does the right thing here:

```
$ (sleep 3; echo foo) | wasirun stdio.wasm                  
waiting for stdin...
foo
```

**Environment (please complete the relevant information):**

```
$ gotip version
go version devel go1.21-13529cc Tue Jun 27 01:09:18 2023 +0000 darwin/arm64
$ uname -a
Darwin MBP.local 22.5.0 Darwin Kernel Version 22.5.0: Thu Jun  8 22:22:20 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T6000 arm64
$ wazero version
v1.2.1
```

Unfortunately I think we need to reopen this issue - after upgrading the Go runners to wazero 1.3.0 we're still seeing errors:

```
##### Test execution environment.
# GOARCH: wasm
# CPU: 
# GOOS: wasip1
# OS Version: UNKNOWN: error determining OS version: unable to determine OS version: unsupported operation

##### ../test
--- FAIL: Test (0.04s)
    --- FAIL: Test/fixedbugs/bug449.go (1.90s)
        testdir_test.go:141: exit status 1
            # command-line-arguments
            ../tmp/Testfixedbugsbug449.go1301620101/001/tmp__.go:3072:6: syntax error: unexpected package, expected name
            ../tmp/Testfixedbugsbug449.go1301620101/001/tmp__.go:6143:6: syntax error: unexpected package, expected name
            ../tmp/Testfixedbugsbug449.go1301620101/001/tmp__.go:9214:6: syntax error: unexpected package, expected name
            
    --- FAIL: Test/clearfat.go (1.78s)
        testdir_test.go:141: exit status 1
            # command-line-arguments
            ../tmp/Testclearfat.go2538136617/001/tmp__.go:4229:13: syntax error: unexpected main at end of statement
            ../tmp/Testclearfat.go2538136617/001/tmp__.go:4234:6: syntax error: unexpected poison1, expected (
            ../tmp/Testclearfat.go2538136617/001/tmp__.go:4236:2: syntax error: unexpected var at end of statement
            ../tmp/Testclearfat.go2538136617/001/tmp__.go:5298:15: exponent has no digits
            ../tmp/Testclearfat.go2538136617/001/tmp__.go:5298:15: syntax error: unexpected ackage at end of statement
            ../tmp/Testclearfat.go2538136617/001/tmp__.go:5303:6: syntax error: unexpected poison1, expected (
            ../tmp/Testclearfat.go2538136617/001/tmp__.go:5305:2: syntax error: unexpected var at end of statement
            ../tmp/Testclearfat.go2538136617/001/tmp__.go:5311:6: syntax error: unexpected clearfat1, expected (
            ../tmp/Testclearfat.go2538136617/001/tmp__.go:5312:2: syntax error: unexpected var at end of statement
            ../tmp/Testclearfat.go2538136617/001/tmp__.go:6367:15: exponent has no digits
            ../tmp/Testclearfat.go2538136617/001/tmp__.go:6367:15: too many errors
```

This is most easily reproduced using the Go testdir tests. They can be run from the [Go source repo](https://go.googlesource.com/go):

1. Build the compiler:
   ```shell
   $ cd src && ./make.bash
   ```
2. Install wazero:
   ```shell
   $ go install github.com/tetratelabs/wazero/cmd/wazero@latest
   ```
3. Run testdir tests (from the `src/` directory):
   ```shell
   $ GOOS=wasip1 GOARCH=wasm PATH=$PATH:$(pwd)/../misc/wasm TMPDIR=$(pwd)/../tmp ../bin/go tool dist test -run 'cmd/internal/testdir:0_1'
   ```

--- a/imports/wasi_snapshot_preview1/testdata/gotip/wasi.go
+++ b/imports/wasi_snapshot_preview1/testdata/gotip/wasi.go
@@ -1,6 +1,7 @@
 package main
 
 import (
+	"bytes"
 	"errors"
 	"fmt"
 	"io"
@@ -8,6 +9,8 @@ import (
 	"net"
 	"net/http"
 	"os"
+	"strconv"
+	"strings"
 	"sync"
 	"syscall"
 	"time"
@@ -53,6 +56,8 @@ func main() {
 		}
 	case "stdout":
 		mainStdout()
+	case "largestdout":
+		mainLargeStdout()
 	}
 }
 
@@ -259,3 +264,46 @@ func mainStdin() error {
 func mainStdout() {
 	os.Stdout.WriteString("test")
 }
+
+func mainLargeStdout() {
+	const ntest = 1024
+
+	var decls, calls bytes.Buffer
+
+	for i := 1; i <= ntest; i++ {
+		s := strconv.Itoa(i)
+		decls.WriteString(strings.Replace(decl, "$", s, -1))
+		calls.WriteString(strings.Replace("call(test$)\n\t", "$", s, -1))
+	}
+
+	program = strings.Replace(program, "$DECLS", decls.String(), 1)
+	program = strings.Replace(program, "$CALLS", calls.String(), 1)
+	fmt.Print(program)
+}
+
+var program = `package main
+
+var count int
+
+func call(f func() bool) {
+	if f() {
+		count++
+	}
+}
+
+$DECLS
+
+func main() {
+	$CALLS
+	if count != 0 {
+		println("failed", count, "case(s)")
+	}
+}
+`
+
+const decl = `
+type T$ [$]uint8
+func test$() bool {
+	v := T${1}
+	return v == [$]uint8{2} || v != [$]uint8{1}
+}`

--- a/imports/wasi_snapshot_preview1/wasi_stdlib_test.go
+++ b/imports/wasi_snapshot_preview1/wasi_stdlib_test.go
@@ -8,6 +8,7 @@ import (
 	"net"
 	"net/http"
 	"os"
+	"os/exec"
 	"path"
 	"runtime"
 	"sort"
@@ -267,6 +268,59 @@ func compileAndRunWithPreStart(t *testing.T, ctx context.Context, config wazero.
 
 	console = consoleBuf.String()
 	return
+}
+
+// compileAndRunForked executes the test case with the wazero runtime in a separate process, and waits for it to terminate.
+//
+// Stdout is captured to a buffer, and stderr is dumped to os.Stderr.
+// It returns the capture buffer and boolean; the boolean is true if we are running in the outer process.
+// If it is false, it means we are running in the subprocess; all the verification logic should be handled
+// when the boolean is true.
+//
+// A typical usage pattern will be:
+//
+//	if buf, hasRun := compileAndRunForked(...); hasRun {
+//		validateContents(buf)
+//	}
+func compileAndRunForked(t *testing.T, ctx context.Context, config wazero.ModuleConfig, tname string, bin []byte) ([]byte, bool) {
+	var buf bytes.Buffer
+	// We use the technique described in https://go.dev/talks/2014/testing.slide#23
+	// We check if we are running forked by ensuring that a "magic" environment variable is set.
+	if os.Getenv("_TEST_FORKED") != "1" {
+		// If said variable is not set, then we need to exec this same executable, specifying the name of test.
+		// We could use t.Name(), but because t may be arbitrarily nested, it is better to get the name of the test
+		// from a parameter.
+		cmd := exec.Command(os.Args[0], "-test.run", tname)
+		cmd.Stdout = &buf
+		cmd.Stderr = os.Stderr
+		cmd.Env = append(os.Environ(), "_TEST_FORKED=1")
+		err := cmd.Run()
+		if e, ok := err.(*exec.ExitError); ok && !e.Success() {
+			require.NoError(t, e, "The test quit with an error code: %v\n", e)
+		}
+		res := buf.Bytes()
+		// This is a test, so in case of success, it will include the "PASS\n" string:
+		// we remove that from the output, to return a clean stdout.
+		return res[0 : len(res)-len("PASS\n")], true
+	}
+
+	r := wazero.NewRuntime(ctx)
+	defer r.Close(ctx)
+
+	_, err := wasi_snapshot_preview1.Instantiate(ctx, r)
+	require.NoError(t, err)
+
+	mod, err := r.InstantiateWithConfig(ctx, bin, config.
+		WithStartFunctions()) // clear
+	require.NoError(t, err)
+
+	_, err = mod.ExportedFunction("_start").Call(ctx)
+	if exitErr, ok := err.(*sys.ExitError); ok {
+		require.Zero(t, exitErr.ExitCode())
+	} else {
+		require.NoError(t, err)
+	}
+	return nil, false
 }
 
 func Test_Poll(t *testing.T) {
@@ -536,4 +590,50 @@ func testStdin(t *testing.T, bin []byte) {
 	_, _ = stdoutReader.Read(buf)
 	require.Equal(t, "foo", string(buf))
 	<-ch
+}
+
+func Test_LargeStdout(t *testing.T) {
+	toolchains := map[string][]byte{}
+	if wasmGotip != nil {
+		toolchains["gotip"] = wasmGotip
+	}
+
+	for toolchain, bin := range toolchains {
+		toolchain := toolchain
+		bin := bin
+		name := t.Name()
+		t.Run(toolchain, func(t *testing.T) {
+			testLargeStdout(t, name, bin)
+		})
+	}
+}
+
+func testLargeStdout(t *testing.T, tname string, bin []byte) {
+	// This test dumps a large Go source file to stdout. The generated result
+	// should be valid code, otherwise it means that stdout is corrupted.
+	//
+	// The error conditions are more easily reproduced by executing in a subprocess
+	// and capturing its stdout.
+	if buf, hasRun := compileAndRunForked(t, testCtx, wazero.NewModuleConfig().
+		WithArgs("wasi", "largestdout").
+		WithStdout(os.Stdout), tname, bin); hasRun {
+
+		tempDir := t.TempDir()
+		temp, err := os.Create(joinPath(tempDir, "out.go"))
+		require.NoError(t, err)
+		defer temp.Close()
+
+		require.NoError(t, err)
+		_, _ = temp.Write(buf)
+		_ = temp.Close()
+
+		gotipBin, err := findGotipBin()
+		require.NoError(t, err)
+
+		cmd := exec.CommandContext(testCtx, gotipBin, "build", "-o",
+			joinPath(tempDir, "outbin"), temp.Name())
+		require.NoError(t, err)
+		output, err := cmd.CombinedOutput()
+		require.NoError(t, err, string(output))
+	}
 }

--- a/internal/sysfs/file_test.go
+++ b/internal/sysfs/file_test.go
@@ -92,6 +92,37 @@ func TestReadFdNonblock(t *testing.T) {
 	require.EqualErrno(t, experimentalsys.EAGAIN, e)
 }
 
+func TestWriteFdNonblock(t *testing.T) {
+	// Test using os.Pipe as it is known to support non-blocking reads.
+	r, w, err := os.Pipe()
+	require.NoError(t, err)
+	defer r.Close()
+	defer w.Close()
+
+	fd := w.Fd()
+	err = setNonblock(fd, true)
+
+	require.NoError(t, err)
+
+	// Create a buffer (the content is not relevant)
+	buf := make([]byte, 1024)
+	// Write to the file until the pipe buffer gets filled up.
+	numWrites := 100
+	for i := 0; i < numWrites; i++ {
+		_, e := writeFd(fd, buf)
+		if e != 0 {
+			if runtime.GOOS == "windows" {
+				// This is currently not supported on Windows.
+				require.EqualErrno(t, experimentalsys.ENOSYS, e)
+			} else {
+				require.EqualErrno(t, experimentalsys.EAGAIN, e)
+			}
+			return
+		}
+	}
+	t.Fatal("writeFd should return EAGAIN at some point")
+}
+
 func TestFileSetAppend(t *testing.T) {
 	tmpDir := t.TempDir()
 

--- a/internal/sysfs/file_unix.go
+++ b/internal/sysfs/file_unix.go
@@ -8,7 +8,10 @@ import (
 	"github.com/tetratelabs/wazero/experimental/sys"
 )
 
-const nonBlockingFileIoSupported = true
+const (
+	nonBlockingFileReadSupported  = true
+	nonBlockingFileWriteSupported = true
+)
 
 // readFd exposes syscall.Read.
 func readFd(fd uintptr, buf []byte) (int, sys.Errno) {
@@ -16,6 +19,16 @@ func readFd(fd uintptr, buf []byte) (int, sys.Errno) {
 		return 0, 0 // Short-circuit 0-len reads.
 	}
 	n, err := syscall.Read(int(fd), buf)
+	errno := sys.UnwrapOSError(err)
+	return n, errno
+}
+
+// writeFd exposes syscall.Write.
+func writeFd(fd uintptr, buf []byte) (int, sys.Errno) {
+	if len(buf) == 0 {
+		return 0, 0 // Short-circuit 0-len writes.
+	}
+	n, err := syscall.Write(int(fd), buf)
 	errno := sys.UnwrapOSError(err)
 	return n, errno
 }

--- a/internal/sysfs/file_unsupported.go
+++ b/internal/sysfs/file_unsupported.go
@@ -2,11 +2,19 @@
 
 package sysfs
 
-import "syscall"
+import "github.com/tetratelabs/wazero/experimental/sys"
 
-const nonBlockingFileIoSupported = false
+const (
+	nonBlockingFileReadSupported  = false
+	nonBlockingFileWriteSupported = false
+)
 
 // readFd returns ENOSYS on unsupported platforms.
-func readFd(fd uintptr, buf []byte) (int, syscall.Errno) {
-	return -1, syscall.ENOSYS
+func readFd(fd uintptr, buf []byte) (int, sys.Errno) {
+	return -1, sys.ENOSYS
+}
+
+// writeFd returns ENOSYS on unsupported platforms.
+func writeFd(fd uintptr, buf []byte) (int, sys.Errno) {
+	return -1, sys.ENOSYS
 }

--- a/internal/sysfs/file_windows.go
+++ b/internal/sysfs/file_windows.go
@@ -7,7 +7,10 @@ import (
 	"github.com/tetratelabs/wazero/experimental/sys"
 )
 
-const nonBlockingFileIoSupported = true
+const (
+	nonBlockingFileReadSupported  = true
+	nonBlockingFileWriteSupported = false
+)
 
 var kernel32 = syscall.NewLazyDLL("kernel32.dll")
 
@@ -55,4 +58,8 @@ func peekNamedPipe(handle syscall.Handle) (uint32, error) {
 		return totalBytesAvail, nil
 	}
 	return totalBytesAvail, err
+}
+
+func writeFd(fd uintptr, buf []byte) (int, sys.Errno) {
+	return -1, sys.ENOSYS
 }

--- a/internal/sysfs/osfile.go
+++ b/internal/sysfs/osfile.go
@@ -145,7 +145,7 @@ func (f *osFile) Read(buf []byte) (n int, errno experimentalsys.Errno) {
 	if len(buf) == 0 {
 		return 0, 0 // Short-circuit 0-len reads.
 	}
-	if nonBlockingFileIoSupported && f.IsNonblock() {
+	if nonBlockingFileReadSupported && f.IsNonblock() {
 		n, errno = readFd(f.fd, buf)
 	} else {
 		n, errno = read(f.file, buf)
@@ -214,7 +214,12 @@ func (f *osFile) Readdir(n int) (dirents []fsapi.Dirent, errno experimentalsys.E
 
 // Write implements the same method as documented on fsapi.File
 func (f *osFile) Write(buf []byte) (n int, errno experimentalsys.Errno) {
-	if n, errno = write(f.file, buf); errno != 0 {
+	if len(buf) == 0 {
+		return 0, 0 // Short-circuit 0-len writes.
+	}
+	if nonBlockingFileWriteSupported && f.IsNonblock() {
+		n, errno = writeFd(f.fd, buf)
+	} else if n, errno = write(f.file, buf); errno != 0 {
 		// Defer validation overhead until we've already had an error.
 		errno = fileError(f, f.closed, errno)
 	}

